### ISSUE TITLE
GraphQL freezes when PP new content

### ISSUE THREAD

**Issue by wezell** [dotCMS/core#19725]
When new content and content types are pushed, we invalidate the graphql schema numerous times.  This causes graphql queries against the production environment to hang until the schema as been rebuilt, which can take a long time.  I think we need a few things:

1. not replace the existing schema until the new one has been built.
2. run the schema cache flush in a commit listener so that it only gets rebuilt once
....

This is an example of the invalidations in the logs

```
[18/12/20 09:44:19:169 EST] INFO util.LogTimeUtil: Call for class: com.dotcms.graphql.business.GraphqlAPIImpl#generateSchema, duration:120023 millis
[18/12/20 09:44:19:226 EST] INFO util.LogTimeUtil: Call for class: com.dotcms.graphql.business.GraphqlAPIImpl#generateSchema, duration:120578 millis
[18/12/20 09:44:19:472 EST] INFO util.LogTimeUtil: Call for class: com.dotcms.graphql.business.GraphqlAPIImpl#generateSchema, duration:123986 millis
[18/12/20 09:44:19:620 EST] INFO util.LogTimeUtil: Call for class: com.dotcms.graphql.business.GraphqlAPIImpl#generateSchema, duration:121361 millis
[18/12/20 09:44:37:961 EST] INFO util.LogTimeUtil: Call for class: com.dotcms.graphql.business.GraphqlAPIImpl#generateSchema, duration:67960 millis
[18/12/20 09:44:42:874 EST] INFO util.LogTimeUtil: Call for class: com.dotcms.graphql.business.GraphqlAPIImpl#generateSchema, duration:72397 millis
```

**Comment by yolabingo** [comment]
cloud support ticket https://dotcms.zendesk.com/agent/tickets/103210

**Comment by nollymar** [comment]
Passed internal qa: 


```
16:28:22.240  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.handler.ContentTypeHandler : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Content Type publish success, ID: 91812c8b-0441-4139-8d4d-7423cfb0e979, Name: Destination
16:28:22.240  INFO  util.SecurityLogger - class com.dotmarketing.portlets.workflows.business.WorkflowAPIImpl : Saving schemes: [WorkflowScheme [id=d61a59e1-a49c-46f2-a929-db2b4bfa88b2, name=System Workflow, archived=false, mandatory=false]], to the content type: com.dotmarketing.portlets.structure.model.Structure@7c9e993b[name=Destination,description=Travel destinations landing pages,defaultStructure=false,reviewInterval=<null>,reviewerRole=<null>,pagedetail=<null>,structureType=5,fixed=false,system=false,velocityVarName=Destination,urlMapPattern=<null>,host=48190c8c-42c4-46af-8d1a-0cd5db894797,folder=SYSTEM_FOLDER,publishDateVar=<null>,expireDateVar=<null>,modDate=Tue Dec 29 15:54:45 EST 2020,iDate=Tue Aug 27 15:00:26 EST 2019,type=structure,owner=<null>,inode=91812c8b-0441-4139-8d4d-7423cfb0e979,identifier=91812c8b-0441-4139-8d4d-7423cfb0e979] -- ip:ukn,user:null
16:28:22.244  INFO  caffine.CaffineCache - ***	 Building Cache : workflowstepcache, size:1000,Concurrency:32
16:28:22.244  INFO  caffine.CaffineCache - ***	 Building Cache : workflowstepcache, size:1000,Concurrency:32
16:28:22.257  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.handler.RelationshipHandler : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Relationship publish (new) success, Inode: ea9b4811-5eac-4c81-8dde-6ae25d0a590f, Name: null location
16:28:22.260  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.handler.RelationshipHandler : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Relationship publish (new) success, Inode: 3f88bdaa-c53a-4176-9ac0-ea459df94351, Name: null activities
16:28:22.277  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.handler.LanguageVariablesHandler : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Language File publish success, Name: cms_language_en.properties
16:28:22.277  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.handler.LanguageVariablesHandler : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Language File publish success, Name: cms_language_en_US.properties
16:28:22.278  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.handler.LanguageVariablesHandler : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Language File publish success, Name: cms_language_es.properties
16:28:22.278  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.handler.LanguageVariablesHandler : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Language File publish success, Name: cms_language_es_ES.properties
16:28:22.279  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.handler.LanguageVariablesHandler : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Language File publish success, Name: cms_language_fr.properties
16:28:22.279  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.handler.LanguageVariablesHandler : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Language File publish success, Name: cms_language_fr_FR.properties
16:28:22.279  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.handler.LanguageVariablesHandler : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Language File publish success, Name: cms_language_zh.properties
16:28:22.280  INFO  util.PushPublishLogger - class com.dotcms.enterprise.publishing.remote.handler.LanguageVariablesHandler : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Language File publish success, Name: cms_language_zh_CN.properties
16:28:22.305  INFO  caffine.CaffineCache - ***	 Building Cache : hostaliascache, size:1000,Concurrency:32
16:28:22.305  INFO  caffine.CaffineCache - ***	 Building Cache : hostaliascache, size:1000,Concurrency:32
16:28:22.311  INFO  util.PushPublishLogger - class com.dotcms.rest.PublishThread : [BundleID: 01ETR6DJC581F0D3TAXTFG4F8X.tar.gz] Finished bundle publish process
16:28:22.311  INFO  rest.PublishThread - Finished bundle publish process
16:28:22.340  INFO  caffine.CaffineCache - ***	 Building Cache : workflowstepcache, size:1000,Concurrency:32
16:28:22.342  INFO  caffine.CaffineCache - ***	 Building Cache : workflowstepcache, size:1000,Concurrency:32
16:28:22.422  INFO  caffine.CaffineCache - ***	 Building Cache : workflowtaskcache, size:10000,Concurrency:32
16:28:22.484  INFO  reindex.BulkProcessorListener - -----------
16:28:22.484  INFO  reindex.BulkProcessorListener - Reindexing Server #  : 1 of 1
16:28:22.484  INFO  reindex.BulkProcessorListener - Total Indexed        : 0
16:28:22.484  INFO  reindex.BulkProcessorListener - ReindexEntries found : 3
16:28:22.484  INFO  reindex.BulkProcessorListener - BulkRequests created : 6
16:28:22.484  INFO  reindex.BulkProcessorListener - -----------
16:28:25.408  INFO  caffine.CaffineCache - ***	 Building Cache : hostaliascache, size:1000,Concurrency:32
16:28:25.408  INFO  caffine.CaffineCache - ***	 Building Cache : hostaliascache, size:1000,Concurrency:32
16:28:25.411  INFO  caffine.CaffineCache - ***	 Building Cache : workflowstepcache, size:1000,Concurrency:32
16:28:25.412  INFO  caffine.CaffineCache - ***	 Building Cache : workflowstepcache, size:1000,Concurrency:32
16:28:27.410  INFO  concurrent.Debouncer - Debouncing : invalidateGraphSchema after 5000 MILLISECONDS
16:28:27.410  INFO  caffine.CaffineCache - ***	 Building Cache : graphqlschemacache, size:1000,Concurrency:32
```

**Comment by bryanboza** [comment]
Fixed, tested after the last changes and now we are invalidating the schema just once,  tested on release 21.01 // Postgres

**Comment by jcastro-dotcms** [comment]
The official code changes have been included in the dotCMS 5.3.8.4 LTS release only. The 5.2.8.4 LTS release is several commits behind, which made the code back-port not possible.
